otter/convergence/selfheal.py

```diff
@@ -0,0 +1,193 @@
+    @inlineCallbacks
+    def _perform(self, config_func, time_range):
+        """
+        Get groups to converge and setup scheduled calls to trigger convergence
+        on each of them within time_range
+        """
+        groups = yield perform(self.disp,
+                               get_groups_to_converge(config_func))
+        active = self._cancel_scheduled_calls()
+        if active:
+            # This should never happen
+            self.log.err(RuntimeError("self-heal-calls-err"),
+                         "self-heal-calls-err", active=active)
+        if not groups:
+            returnValue(None)
+        wait_time = float(time_range) / len(groups)
+        for i, group in enumerate(groups):
+            self.calls.append(
+                self.clock.callLater(
+                    i * wait_time, perform, self.disp,
+                    check_and_trigger(group["tenantId"], group["groupId"]))
+            )
+
+    @inlineCallbacks
+    def _converge_all(self, config_func, time_range):
+        """
+        Setup convergence triggering on all groups by acquiring the lock and
+        calling _perform
+        """
+        if self.kz_client.state != KazooState.CONNECTED:
+            self.log.err(RuntimeError("self-heal-kz-state"),
+                         "self-heal-kz-state", state=self.kz_client.state)
+            returnValue(None)
+        # The reason why it checks everytime by talking to ZK is because
+        # we could've lost the lock if there where connection changes between
+        # subsequent intervals and it is not clear how state change to
+        # SUSPENDED should be handled
+        if (yield is_lock_acquired(self.disp, self.lock)):
+            yield self._perform(config_func, time_range)
+        else:
+            try:
+                yield self.lock.acquire(True, 0.1)
+                self.log.msg("self-heal-lock-acquired")
+                yield self._perform(config_func, time_range)
+            except LockTimeout:
+                # expected. Nothing to do here. Will try on next interval
+                pass
+
+
+def get_groups_to_converge(config_func):
+    """
+    Get all tenant's all groups that needs convergence triggering
+    """
+    eff = Effect(GetAllValidGroups())
+    eff = eff.on(
+        filter(lambda g: tenant_is_enabled(g["tenantId"], config_func)))
+    return eff.on(list)
+
+
+@do
+def check_and_trigger(tenant_id, group_id):
+    """
+    Trigger convergence on given group if it is ACTIVE and not paused
+    """
+    group, info = yield Effect(
+        GetScalingGroupInfo(tenant_id=tenant_id, group_id=group_id))
+    state = info["state"]
+
+    if state.status == ScalingGroupStatus.ACTIVE and (not state.paused):
+        yield with_log(
+            trigger_convergence(tenant_id, group_id),
+            tenant_id=tenant_id, scaling_group_id=group_id)
+
+
+def is_lock_acquired(dispatcher, lock):
+    """
+    Does given lock object has acquired the lock? It does this by getting
```

Comment:
Grammar: this sentence is a mess 😄 .  I think you mean something like: "Is the given lock object currently acquired by this worker?"